registry/storage/catalog.go

```diff
@@ -81,78 +79,95 @@ func (reg *registry) Remove(ctx context.Context, name reference.Named) error {
 }
 
-// compareReplaceInline modifies runtime.cmpstring to replace old with new
-// during a byte-wise comparison.
-func compareReplaceInline(s1, s2 string, old, new byte) int {
-	// TODO(stevvooe): We are missing an optimization when the s1 and s2 have
-	// the exact same slice header. It will make the code unsafe but can
-	// provide some extra performance.
+// getRepositoriesRec recurse through all folders it the `lookPath`,
+// there it will try to find repositories. See getRepositories for more.
+func (reg *registry) getRepositoriesRec(ctx context.Context, root, lookPath, last string, fn func(repoPath string) error) error {
+	// ensure that the current path is a dir, otherwise we just return
+	if f, err := reg.blobStore.driver.Stat(ctx, lookPath); err != nil || !f.IsDir() {
+		if err != nil {
+			return err
+		}
+		return nil
+	}
 
-	l := len(s1)
-	if len(s2) < l {
-		l = len(s2)
+	// get children in the current path
+	children, err := reg.blobStore.driver.List(ctx, lookPath)
+	if err != nil {
+		return err
 	}
 
-	for i := 0; i < l; i++ {
-		c1, c2 := s1[i], s2[i]
-		if c1 == old {
-			c1 = new
-		}
+	// sort this, so that it will be added in the correct order
+	sort.Strings(children)
 
-		if c2 == old {
-			c2 = new
-		}
+	if last != "" {
+		splitLasts := strings.Split(last, "/")
 
-		if c1 < c2 {
-			return -1
+		// call the next iteration of getRepositoriesRec if any, but
+		// exclude the current one.
+		if len(splitLasts) > 1 {
+			if err := reg.getRepositoriesRec(ctx, root, lookPath+"/"+splitLasts[0], strings.Join(splitLasts[1:], "/"), fn); err != nil {
+				return err
+			}
 		}
 
-		if c1 > c2 {
-			return +1
+		// find current last path in our children
+		n := sort.SearchStrings(children, lookPath+"/"+splitLasts[0])
+		if n == len(children) || children[n] != lookPath+"/"+splitLasts[0] {
+			return errors.New("the provided 'last' repositories does not exists")
 		}
-	}
-
-	if len(s1) < len(s2) {
-		return -1
-	}
 
-	if len(s1) > len(s2) {
-		return +1
+		// if this is not a final `last` (there are more `/` left)
+		// then exclude the current index, else include it
+		if len(splitLasts) > 1 {
+			children = children[n+1:]
+		} else {
+			children = children[n:]
+		}
 	}
 
-	return 0
-}
+	for _, child := range children {
+		_, file := path.Split(child)
 
-// handleRepository calls function fn with a repository path if fileInfo
-// has a path of a repository under root and that it is lexographically
-// after last. Otherwise, it will return ErrSkipDir. This should be used
-// with Walk to do handling with repositories in a storage.
-func handleRepository(fileInfo driver.FileInfo, root, last string, fn func(repoPath string) error) error {
-	filePath := fileInfo.Path()
-
-	// lop the base path off
-	repo := filePath[len(root)+1:]
-
-	_, file := path.Split(repo)
-	if file == "_manifests" {
-		repo = strings.TrimSuffix(repo, "/_manifests")
-		if lessPath(last, repo) {
-			if err := fn(repo); err != nil {
+		if file == "_manifest" {
```

Comment:
This is a typo causing the tests to fail. It should be **plural**: `"_manifests"`